cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,431 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * A Spark program for generating reports, querying for report statuses, and reading reports.
+ */
+public class ReportGenerationSpark extends AbstractExtendedSpark implements JavaSparkMain {
+  private static final Logger LOG = LoggerFactory.getLogger(ReportGenerationSpark.class);
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Filter.class, new FilterDeserializer())
+    .create();
+
+  @Override
+  protected void configure() {
+    setMainClass(ReportGenerationSpark.class);
+    addHandlers(new ReportSparkHandler());
+  }
+
+  @Override
+  public void run(JavaSparkExecutionContext sec) throws Exception {
+    JavaSparkContext jsc = new JavaSparkContext();
+  }
+
+  /**
+   * A {@link SparkHttpServiceHandler} generating reports, querying for report statuses, and reading reports.
+   */
+  public static final class ReportSparkHandler extends AbstractSparkHttpServiceHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReportSparkHandler.class);
+    private static final Type REPORT_GENERATION_REQUEST_TYPE = new TypeToken<ReportGenerationRequest>() {
+    }.getType();
+    private static final String DEFAULT_LIMIT = "10000";
+    private SparkSession sparkSession;
+
+    public static final String READ_LIMIT = "readLimit";
+    public static final String THREAD_POOL_SIZE = "poolSize";
+    public static final String START_FILE = "_START";
+    public static final String REPORT_DIR = "reports";
+    public static final String COUNT_FILE = "COUNT";
+    public static final String SUCCESS_FILE = "_SUCCESS";
+    public static final String FAILURE_FILE = "_FAILURE";
+
+    private int readLimit;
+    private ExecutorService reportGenerationExecutor;
+
+    @Override
+    public void initialize(SparkHttpServiceContext context) throws Exception {
+      super.initialize(context);
+      sparkSession = new SQLContext(getContext().getSparkContext()).sparkSession();
+      String readLimitString = context.getRuntimeArguments().get(READ_LIMIT);
+      readLimit = readLimitString == null ? 10000 : Integer.valueOf(readLimitString);
+      String threadPoolSizeString = context.getRuntimeArguments().get(THREAD_POOL_SIZE);
+      reportGenerationExecutor =
```

Comment:
Why will this get shutdown? Also, beware that in handler, the `initialize` and `destroy` methods will get invoked for different handler instance, and there is one handler instance per http worker thread. You might want to use a singleton executor instead, with dynamic threadpool with core size = 0 and daemon thread.